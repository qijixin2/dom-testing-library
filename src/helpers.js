const globalObj = typeof window === 'undefined' ? global : window
// Constant node.nodeType for text nodes, see:
// https://developer.mozilla.org/en-US/docs/Web/API/Node/nodeType#Node_type_constants
const TEXT_NODE = 3

// Currently this fn only supports jest timers, but it could support other test runners in the future.
function runWithRealTimers(callback) {
  return _runWithRealTimers(callback).callbackReturnValue
}

function _runWithRealTimers(callback) {
  const timerAPI = {
    clearImmediate,
    clearInterval,
    clearTimeout,
    setImmediate,
    setInterval,
    setTimeout,
  }

  // istanbul ignore else
  if (typeof jest !== 'undefined') {
    jest.useRealTimers()
  }

  const callbackReturnValue = callback()

  const usedJestFakeTimers = Object.entries(timerAPI).some(
    ([name, func]) => func !== globalObj[name],
  )

  if (usedJestFakeTimers) {
    jest.useFakeTimers(timerAPI.setTimeout?.clock ? 'modern' : 'legacy')
  }

  return {
    callbackReturnValue,
    usedJestFakeTimers,
  }
}

const jestFakeTimersAreEnabled = () =>
  Boolean(_runWithRealTimers(() => {}).usedJestFakeTimers)

// we only run our tests in node, and setImmediate is supported in node.
// istanbul ignore next
function setImmediatePolyfill(fn) {
  return globalObj.setTimeout(fn, 0)
}

function getTimeFunctions() {
  // istanbul ignore next
  return {
    clearTimeoutFn: globalObj.clearTimeout,
    setImmediateFn: globalObj.setImmediate || setImmediatePolyfill,
    setTimeoutFn: globalObj.setTimeout,
  }
}

const {clearTimeoutFn, setImmediateFn, setTimeoutFn} = runWithRealTimers(
  getTimeFunctions,
)

function getDocument() {
  /* istanbul ignore if */
  if (typeof window === 'undefined') {
    throw new Error('Could not find default container')
  }
  return window.document
}
function getWindowFromNode(node) {
  if (node.defaultView) {
    // node is document
    return node.defaultView
  } else if (node.ownerDocument && node.ownerDocument.defaultView) {
    // node is a DOM node
    return node.ownerDocument.defaultView
  } else if (node.window) {
    // node is window
    return node.window
  } else if (node.then instanceof Function) {
    throw new Error(
      `It looks like you passed a Promise object instead of a DOM node. Did you do something like \`fireEvent.click(screen.findBy...\` when you meant to use a \`getBy\` query \`fireEvent.click(screen.getBy...\`, or await the findBy query \`fireEvent.click(await screen.findBy...\`?`,
    )
  } else {
    // The user passed something unusual to a calling function
    throw new Error(
      `Unable to find the "window" object for the given node. Please file an issue with the code that's causing you to see this error: https://github.com/testing-library/dom-testing-library/issues/new`,
    )
  }
}

function checkContainerType(container) {
  if (
    !container ||
    !(typeof container.querySelector === 'function') ||
    !(typeof container.querySelectorAll === 'function')
  ) {
    throw new TypeError(
      `Expected container to be an Element, a Document or a DocumentFragment but got ${getTypeName(
        container,
      )}.`,
    )
  }

  function getTypeName(object) {
    if (typeof object === 'object') {
      return object === null ? 'null' : object.constructor.name
    }
    return typeof object
  }
}

export {
  getWindowFromNode,
  getDocument,
  clearTimeoutFn as clearTimeout,
  setImmediateFn as setImmediate,
  setTimeoutFn as setTimeout,
  runWithRealTimers,
  checkContainerType,
  jestFakeTimersAreEnabled,
  TEXT_NODE,
}
