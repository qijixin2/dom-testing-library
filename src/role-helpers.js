import {elementRoles} from 'aria-query'
import {prettyDOM} from './pretty-dom'

const elementRoleList = buildElementRoleList(elementRoles)

function getImplicitAriaRoles(currentNode) {
  for (const {selector, roles} of elementRoleList) {
    if (currentNode.matches(selector)) {
      return [...roles]
    }
  }

  return []
}

function buildElementRoleList(elementRolesMap) {
  function makeElementSelector({name, attributes = []}) {
    return `${name}${attributes
      .map(({name: attributeName, value}) =>
        value ? `[${attributeName}=${value}]` : `[${attributeName}]`,
      )
      .join('')}`
  }

  function getSelectorSpecificity({attributes = []}) {
    return attributes.length
  }

  function bySelectorSpecificity(
    {specificity: leftSpecificity},
    {specificity: rightSpecificity},
  ) {
    return rightSpecificity - leftSpecificity
  }

  let result = []

  for (const [element, roles] of elementRolesMap.entries()) {
    result = [
      ...result,
      {
        selector: makeElementSelector(element),
        roles: Array.from(roles),
        specificity: getSelectorSpecificity(element),
      },
    ]
  }

  return result.sort(bySelectorSpecificity)
}

function getRoles(container) {
  function flattenDOM(node) {
    return [
      node,
      ...Array.from(node.children).reduce(
        (acc, child) => [...acc, ...flattenDOM(child)],
        [],
      ),
    ]
  }

  return flattenDOM(container).reduce((acc, node) => {
    const roles = getImplicitAriaRoles(node)

    return roles.reduce(
      (rolesAcc, role) =>
        Array.isArray(rolesAcc[role])
          ? {...rolesAcc, [role]: [...rolesAcc[role], node]}
          : {...rolesAcc, [role]: [node]},
      acc,
    )
  }, {})
}

function prettyRoles(dom) {
  const roles = getRoles(dom)

  return Object.entries(roles)
    .map(([role, elements]) => {
      const delimiterBar = '-'.repeat(50)
      const elementsString = elements
        .map(el => prettyDOM(el.cloneNode(false)))
        .join('\n\n')

      return `${role}:\n\n${elementsString}\n\n${delimiterBar}`
    })
    .join('\n')
}

const logRoles = dom => console.log(prettyRoles(dom))

export {getRoles, logRoles, getImplicitAriaRoles, prettyRoles}

/* eslint no-console:0 */
