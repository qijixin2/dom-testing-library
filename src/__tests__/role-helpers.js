import jestSerializerAnsi from 'jest-serializer-ansi'
import {getRoles, logRoles, getImplicitAriaRoles} from '../role-helpers'
import {render, cleanup} from './helpers/test-utils'

afterEach(cleanup)

function setup() {
  const {getByTestId} = render(`
<section data-testid='a-section'>
  <nav data-testid='a-nav' />
  
  <h1 data-testid='a-h1'>Main Heading</h1>
  <h2 data-testid='a-h2'>Sub Heading</h2>
  <h3 data-testid='a-h3'>Tertiary Heading</h3>

  <article data-testid='a-article'>
    <!-- menuitem is currently deprecated, but is the only 
         tag currently that aria-query returns multiple roles for
         (roles: command, menuitem).
         It's used here in case a future tag also has multiple 
         roles -->
    <menuitem data-testid='a-menuitem-1'>1</menuitem> 
    <menuitem data-testid='a-menuitem-2'>2</menuitem>

    <ul data-testid='a-list'>
      <li data-testid='a-list-item-1'>Item 1</li>
      <li data-testid='a-list-item-2'>Item 2</li>
    </ul>

    <table data-testid='a-table'>
      <tbody data-testid='a-tbody'>
        <tr data-testid='a-row'>
          <td data-testid='a-cell-1'>Cell 1</td>
          <td data-testid='a-cell-2'>Cell 2</td>
          <td data-testid='a-cell-3'>Cell 3</td>
        </tr>
      </tbody>
    </table>

    <form data-testid='a-form'>
      <input type='radio' data-testid='a-radio-1' />
      <input type='radio' data-testid='a-radio-2' />
      <input type='text' data-testid='a-input-1' />
      <input type='text' data-testid='a-input-2' />
      <textarea data-testid='a-textarea'></textarea>
    </form>

    <ul data-testid='b-list'>
      <li data-testid='b-list-item-1'>Item 1</li>
      <li data-testid='b-list-item-2'>Item 2</li>
    </ul>
   </article>
</section>
  `)

  return {
    section: getByTestId('a-section'),
    h1: getByTestId('a-h1'),
    h2: getByTestId('a-h2'),
    h3: getByTestId('a-h3'),
    nav: getByTestId('a-nav'),
    article: getByTestId('a-article'),
    menuItem: getByTestId('a-menuitem-1'),
    menuItem2: getByTestId('a-menuitem-2'),
    aUl: getByTestId('a-list'),
    aLi1: getByTestId('a-list-item-1'),
    aLi2: getByTestId('a-list-item-2'),
    bUl: getByTestId('b-list'),
    bLi1: getByTestId('b-list-item-1'),
    bLi2: getByTestId('b-list-item-2'),
    table: getByTestId('a-table'),
    tbody: getByTestId('a-tbody'),
    tr: getByTestId('a-row'),
    td1: getByTestId('a-cell-1'),
    td2: getByTestId('a-cell-2'),
    td3: getByTestId('a-cell-3'),
    form: getByTestId('a-form'),
    radio: getByTestId('a-radio-1'),
    radio2: getByTestId('a-radio-2'),
    input: getByTestId('a-input-1'),
    input2: getByTestId('a-input-2'),
    textarea: getByTestId('a-textarea'),
  }
}

test('getRoles returns expected roles for various dom nodes', () => {
  const {
    section,
    h1,
    h2,
    h3,
    nav,
    article,
    menuItem,
    menuItem2,
    aUl,
    aLi1,
    aLi2,
    bUl,
    bLi1,
    bLi2,
    table,
    tbody,
    tr,
    td1,
    td2,
    td3,
    form,
    radio,
    radio2,
    input,
    input2,
    textarea,
  } = setup()

  expect(getRoles(section)).toEqual({
    region: [section],
    heading: [h1, h2, h3],
    navigation: [nav],
    radio: [radio, radio2],
    article: [article],
    list: [aUl, bUl],
    listitem: [aLi1, aLi2, bLi1, bLi2],
    table: [table],
    row: [tr],
    cell: [td1, td2, td3],
    form: [form],
    textbox: [input, input2, textarea],
    rowgroup: [tbody],
    command: [menuItem, menuItem2],
    menuitem: [menuItem, menuItem2],
  })
})

test('logRoles logs expected roles for various dom nodes', () => {
  expect.addSnapshotSerializer(jestSerializerAnsi)

  const {section} = setup()
  const output = logRoles(section)

  expect(output).toMatchSnapshot()
})

test('getImplicitAriaRoles returns expected roles for various dom nodes', () => {
  const {section, h1, form, radio, input} = setup()

  expect(getImplicitAriaRoles(section)).toEqual(['region'])
  expect(getImplicitAriaRoles(h1)).toEqual(['heading'])
  expect(getImplicitAriaRoles(form)).toEqual(['form'])
  expect(getImplicitAriaRoles(radio)).toEqual(['radio'])
  expect(getImplicitAriaRoles(input)).toEqual(['textbox'])
})
