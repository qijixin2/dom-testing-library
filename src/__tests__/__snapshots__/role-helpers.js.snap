// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`logRoles logs expected roles for various dom nodes 1`] = `
"region:

<section
  data-testid="a-section"
/>

--------------------------------------------------
navigation:

<nav
  data-testid="a-nav"
/>

--------------------------------------------------
heading:

<h1
  data-testid="a-h1"
/>

<h2
  data-testid="a-h2"
/>

<h3
  data-testid="a-h3"
/>

--------------------------------------------------
article:

<article
  data-testid="a-article"
/>

--------------------------------------------------
command:

<menuitem
  data-testid="a-menuitem-1"
/>

<menuitem
  data-testid="a-menuitem-2"
/>

--------------------------------------------------
menuitem:

<menuitem
  data-testid="a-menuitem-1"
/>

<menuitem
  data-testid="a-menuitem-2"
/>

--------------------------------------------------
list:

<ul
  data-testid="a-list"
/>

<ul
  data-testid="b-list"
/>

--------------------------------------------------
listitem:

<li
  data-testid="a-list-item-1"
/>

<li
  data-testid="a-list-item-2"
/>

<li
  data-testid="b-list-item-1"
/>

<li
  data-testid="b-list-item-2"
/>

--------------------------------------------------
table:

<table
  data-testid="a-table"
/>

--------------------------------------------------
rowgroup:

<tbody
  data-testid="a-tbody"
/>

--------------------------------------------------
row:

<tr
  data-testid="a-row"
/>

--------------------------------------------------
cell:

<td
  data-testid="a-cell-1"
/>

<td
  data-testid="a-cell-2"
/>

<td
  data-testid="a-cell-3"
/>

--------------------------------------------------
form:

<form
  data-testid="a-form"
/>

--------------------------------------------------
radio:

<input
  data-testid="a-radio-1"
  type="radio"
/>

<input
  data-testid="a-radio-2"
  type="radio"
/>

--------------------------------------------------
textbox:

<input
  data-testid="a-input-1"
  type="text"
/>

<input
  data-testid="a-input-2"
  type="text"
/>

<textarea
  data-testid="a-textarea"
/>

--------------------------------------------------"
`;
